apps/tools/scxmldump/main.cpp

```diff
@@ -450,36 +459,83 @@ class EventDump : public Seiscomp::Client::Application {
+			vector<string> journalingPublicIDs;
+
+			// parse -E and -O command-line arguments and collect the publicID's
+
+			vector<string> eventIDs;
 			if ( commandline().hasOption("event") ) {
-				EventParametersPtr ep = new EventParameters;
-				vector<string> eventIDs;
 				Core::split(eventIDs, _eventIDs.c_str(), ",");
-				for ( vector<string>::const_iterator it = eventIDs.begin();
-				      it != eventIDs.end(); ++it ) {
+				for (const string &publicID : eventIDs)
+					journalingPublicIDs.push_back(publicID);
+			}
+
+			vector<string> originIDs;
+			if ( commandline().hasOption("origin") ) {
+				Core::split(originIDs, _originIDs.c_str(), ",");
+				for (const string &publicID : originIDs)
+					journalingPublicIDs.push_back(publicID);
+
+			}
+
+			JournalingPtr jnl = NULL;
+			if ( commandline().hasOption("journal") ) {
+				if ( ! journalingPublicIDs.empty() ) {
+					// create a journal specific to the publicIDs
+					jnl = new Journaling();
+					for (const string &publicID : journalingPublicIDs) {
+						DatabaseIterator it = query()->getJournal(publicID);
+						while ( it.get() ) {
+							jnl->add(JournalEntry::Cast(*it));
+							++it;
+						}
+					}
+				}
+				else {
+					// retrieve entire journal
+					jnl = query()->loadJournaling();
+					if ( !jnl ) {
+						SEISCOMP_ERROR("Journal has not been found");
+						return false;
+					}
+				}
+			}
+
+			EventParametersPtr ep = NULL;
+			if ( ! eventIDs.empty() ) {
+				if ( ! ep)
+					ep = new EventParameters;
+				for (const string &publicID : eventIDs) {
 					EventPtr event = Event::Cast(PublicObjectPtr(
-						query()->getObject(Event::TypeInfo(), *it)));
+						query()->getObject(Event::TypeInfo(), publicID)));
 					if ( event )
 						addEvent(ep.get(), event.get());
 					else
-						SEISCOMP_ERROR("Event with id '%s' has not been found", it->c_str());
+						SEISCOMP_ERROR("Event with id '%s' has not been found", publicID.c_str());
 				}
-				return ep->eventCount() > 0 && write(ep.get());
 			}
 
-			if ( commandline().hasOption("origin") ) {
-				EventParametersPtr ep = new EventParameters;
+			if ( ! originIDs.empty() ) {
+				if ( ! ep)
+					ep = new EventParameters;
 				vector<string> originIDs;
 				Core::split(originIDs, _originIDs.c_str(), ",");
-				for ( vector<string>::const_iterator it = originIDs.begin();
-				      it != originIDs.end(); ++it ) {
+				for (const string &publicID : originIDs) {
 					OriginPtr origin = Origin::Cast(PublicObjectPtr(
-						query()->getObject(Origin::TypeInfo(), *it)));
+						query()->getObject(Origin::TypeInfo(), publicID)));
 					if ( origin )
 						addOrigin(ep.get(), origin.get());
 					else
-						SEISCOMP_ERROR("Origin with id '%s' has not been found", it->c_str());
+						SEISCOMP_ERROR("Origin with id '%s' has not been found", publicID.c_str());
 				}
-				return ep->originCount() > 0 && write(ep.get());
+			}
+
+			if (ep != NULL || jnl != NULL) {
```

Comment:
Spaces and use logical expressions: `if ( ep && jnl ) {`. You can also use `and` instead of `&&` which improves readability.

Reply:
Speaking of readability, "ep != NULL || jnl != NULL" has the best readability IMO, because you immediately know that ep and jn are pointers and not boolean variables. Wasn't it suggested to use "is None" in Python?

OK, "ep || jnl" is more compact and makes sense, but I don't agree that "ep and jnl" (or do you mean "ep or jnl"?) is more readable than "ep && jnl" for C programmers...

Reply:
I mean "or" yes. Sorry for the mistake. Python is not a typed language, C++ is. The "is None" thing is needed to distinguish cases where a number is zero or a string is empty. That is never the case in C++. You know the type of a variable. To be pedantic, `ep` is not a pointer, it is a class. Anyway, we agreed on a style guide long time ago so I would not like to open this issue again: https://docs.gempa.de/seiscomp/current/base/coding-conventions.html#null-pointer